include/rmm/detail/thrust_namespace.h

```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#ifdef THRUST_WRAPPED_NAMESPACE
+
+// Ensure the namespace exist before we import it
+// so that this include can occur before thrust includes
+namespace THRUST_WRAPPED_NAMESPACE {
+    namespace thrust {
+    }
+}
+using namespace THRUST_WRAPPED_NAMESPACE;
```

Comment:
I'm not sure I fully understand all of this. However, it seems to me there's a concern with the `using namespace THRUST_WRAPPED_NAMESPACE;` since that will affect all code following the inclusion of some rmm headers. Presumably people use `THRUST_WRAPPED_NAMESPACE` since they may deal with multiple versions of thrust that are being included in different parts of the code, and I feel like it may be dangerous to possibly pull in the given wrapped thrust in downstream user code, since the original idea was to use the wrapped namespace to keep it separate? That is, it'd be safer to have this `using namespace THRUST_WRAPPED_NAMESPACE` statement inside of `namespace rmm {}`?

Reply:
> Presumably people use THRUST_WRAPPED_NAMESPACE since they may deal with multiple versions of thrust that are being included in different parts of the code, and I feel like it may be dangerous to possibly pull in the given wrapped thrust in downstream user code, since the original idea was to use the wrapped namespace to keep it separate?

Only a single instance of Thrust is possible in a given compilation instance/file. So when compiling a source file that places Thrust in a custom location rmm needs to know what that location is. This change ensures that any call to `thrust::` still works but will be properly managled as desired.

> safer to have this using namespace THRUST_WRAPPED_NAMESPACE statement inside of namespace rmm {}?

Techincally this approach would work, but rmm can't use it. The reason being is that `rmm::device_vector` is a `thrust::device_vector` with a custom allocator. When we import thrust into the rmm namespace we end up with two different definitions of the same class.

Reply:
Or are you saying we bring thrust like this?
```cxx
namespace THRUST_WRAPPED_NAMESPACE {
    namespace thrust {
    }
}

namespace rmm {
using namespace THRUST_WRAPPED_NAMESPACE;
}
```

Reply:
That is what I was thinking, but if it's not supported to have multiple (wrapped) versions of thrust in a single compilation unit anyway, then I suppose it doesn't really make a difference.

Well, it'll still prevent the user that includes an rmm header to write `thrust::device_vector<>` and have it seem to work when they were supposed to write `WRAPPED::thrust_vector<>`. OTOH, I can only assume that some library using a wrapped thrust would have a `using namespace WRAPPED;` in their .cpp files anyway...

Reply:
Okay I updated to use your proposed version, moving out of draft.